Sources/Testing/Parameterization/TypeInfo.swift

```diff
@@ -12,38 +12,125 @@
+    ///   - fullyQualified: The fully-qualified name of the type.
+    ///   - unqualified: The unqualified name of the type.
+    case nameOnly(fullyQualified: String, unqualified: String)
+  }
+
+  /// The kind of type info.
+  private var _kind: _Kind
+
+  /// The complete name of this type, with the names of all referenced types
+  /// fully-qualified by their module names when possible.
+  ///
+  /// The value of this property is equal to ``fullyQualifiedName``, but is
+  /// split into components. For instance, given the following declaration in
+  /// the `Example` module:
+  ///
+  /// ```swift
+  /// struct A {
+  ///   struct B {}
+  /// }
+  /// ```
+  ///
+  /// The value of this property for the type `A.B` would be
+  /// `["Example", "A", "B"]`.
+  public var fullyQualifiedNameComponents: [String] {
+    switch _kind {
+    case let .type(type):
+      nameComponents(of: type)
+    case let .nameOnly(fqn, _):
+      fqn.split(separator: ".").map(String.init)
+    }
+  }
+
   /// The complete name of this type, with the names of all referenced types
   /// fully-qualified by their module names when possible.
-  public var qualifiedName: String
+  ///
+  /// The value of this property is equal to ``fullyQualifiedNameComponents``,
+  /// but is represented as a single string. For instance, given the following
+  /// declaration in the `Example` module:
+  ///
+  /// ```swift
+  /// struct A {
+  ///   struct B {}
+  /// }
+  /// ```
+  ///
+  /// The value of this property for the type `A.B` would be `"Example.A.B"`.
+  public var fullyQualifiedName: String {
+    switch _kind {
+    case let .type(type):
+      Testing.fullyQualifiedName(of: type)
+    case let .nameOnly(fqn, _):
+      fqn
+    }
+  }
 
   /// A simplified name of this type, by leaving the names of all referenced
   /// types unqualified, i.e. without module name prefixes.
-  public var unqualifiedName: String
-
-  init(
-    qualifiedName: String,
-    unqualifiedName: String
-  ) {
-    self.qualifiedName = qualifiedName
-    self.unqualifiedName = unqualifiedName
+  ///
+  /// The value of this property is equal to the name of the type in isolation.
+  /// For instance, given the following declaration in the `Example` module:
+  ///
+  /// ```swift
+  /// struct A {
+  ///   struct B {}
+  /// }
+  /// ```
+  ///
+  /// The value of this property for the type `A.B` would simply be `"B"`.
+  public var unqualifiedName: String {
+    switch _kind {
+    case let .type(type):
+      String(describing: type)
+    case let .nameOnly(_, unqualifiedName):
+      unqualifiedName
+    }
+  }
+
+  /// The described type, if available.
+  ///
+  /// If this instance was created from a type name, or if it was previously
+  /// encoded and decoded, the value of this property is `nil`.
+  public var type: Any.Type? {
+    if case let .type(type) = _kind {
+      return type
+    }
+    return nil
+  }
+
+  init(fullyQualifiedName: String, unqualifiedName: String) {
```

Comment:
As of this PR, this initializer is only being used in the `Decodable` implementation, right?

Reply:
It's also used in our tests. :)